lib/geckodriver.js

```diff
@@ -0,0 +1,180 @@
+      return true;
+    });
+    return version;
+  });
+
+  async start (caps, emitStartingState = true) {
+    this.capabilities = _.cloneDeep(caps.desiredCapabilities || caps.capabilities);
+
+    const args = [`--host=${this.proxyHost}`, `--port=${this.proxyPort}`, '-vv'];
+
+    const startDetector = (stdout) => {
+      return stdout.indexOf('Listening') !== -1;
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    this.geckodriver = SYSTEM_GECKODRIVER;
+    let version = await this.getVersion();
+    log.debug(`Geckodriver version ${version}`);
+    if (compareVersions(version, MIN_SUPPORTED_VERSION, '<')) {
+      log.errorAndThrow(`Only geckodriver ${MIN_SUPPORTED_VERSION} supports Android testing. Current version: ${version}`);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.proc = new SubProcess(this.geckodriver, args);
+      processIsAlive = true;
+
+      this.proc.on('output', (stdout, stderr) => {
+        if (this.verbose) {
+          for (let line of (stdout || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.debug(`[STDOUT] ${line}`);
+          }
+          for (let line of (stderr || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.error(`[STDERR] ${line}`);
+          }
+        }
+      });
+
+      this.proc.on('exit', (code, signal) => {
+        log.info(code, signal);
+        processIsAlive = false;
+
+        if (this.state !== Geckodriver.STATE_STOPPED &&
+          this.state !== Geckodriver.STATE_STOPPING &&
+          this.state !== Geckodriver.STATE_RESTARTING) {
+          let msg = `Geckodriver exited unexpectedly with code ${code}, ` +
+                  `signal ${signal}`;
+          log.error(msg);
+          this.changeState(Geckodriver.STATE_STOPPED);
+        }
+      });
+
+      log.info(`Spawning geckodriver with: ${this.geckodriver} ` +
+              `${args.join(' ')}`);
+      // start subproc and wait for startDetector
+      await this.proc.start(startDetector);
+      this.changeState(Geckodriver.STATE_ONLINE);
+      await this.startSession();
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+      if (processIsAlive) {
+        await this.stop();
+      }
+    }
+  }
+
+  sessionId () {
+    if (this.state !== Geckodriver.STATE_ONLINE) {
+      return null;
+    }
+    return this.jwproxy.sessionId;
+  }
+
+  async stop () {
+    try {
+      await this.proc.stop();
+      this.changeState(Geckodriver.STATE_STOPPED);
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+    }
+  }
+
+  async getStatus () {
+    return await this.jwproxy.command('/status', 'GET');
+  }
+
+  prepareMozCapabilities () {
+    let mozFirefoxOptions = {
+      log: {
+        level: this.logLevel
+      },
+      androidPackage: FIREFOX_PACKAGE,
```

Comment:
do we only want to limit the driver to Firefox?

Reply:
you mean to just Android, right? We can add support for geckodriver in general

Reply:
I mean we could still have package and activity as capabilities, but set these as defaults

Reply:
sounds good, I'll check for `platformName` if it's set and it's android we use these default if they're not sent.